KAFKA-19899: Bump group epoch when member regex subscription transitions from non-empty to empty

AntonVasant opened the pull request:
This PR fixes an issue in
GroupMetadataManager#maybeUpdateRegularExpressions where a member’s
regex subscription transition from non-empty → empty did not trigger a
group epoch bump.  The method previously returned REGEX_UPDATED, which
does not cause consumerGroupHeartbeat to increment the group epoch.

Fix  The patch updates the logic to return: REGEX_UPDATED_AND_RESOLVED

when:

the updated regex subscription text is empty, and  the previous
subscription was non-empty.  This ensures that consumerGroupHeartbeat
correctly bumps the group epoch, keeping the group metadata consistent.

Tests

Several tests were updated to align with the corrected behavior.  Tests
that previously expected no epoch bump were failing, and have now been
adjusted to expect the new, correct logic.

JIRA  https://issues.apache.org/jira/browse/KAFKA-19891

Impact

Fixes coordinator state correctness for regex-subscribing consumer
groups  Ensures group epoch bumps happen for all relevant subscription
transitions  Backward compatible  No public API changes

Reviewers: Sean Quah <squah@confluent.io>, David Jacot
 <djacot@confluent.io>

AntonVasant committed:
KAFKA-19891: Bump group epoch when member regex subscription transitions from non-empty to empty

This patch fixes an issue in
where  was incorrectly set to
when a member’s regex subscription transitions from non-empty to empty.

Because  does not trigger a group epoch bump in
, the group metadata could remain stale. The fix
updates the method to return  in this case,
ensuring the group epoch is correctly incremented.

The patch also updates tests that relied on the previous behavior and were
failing due to the corrected epoch bump logic.

AntonVasant committed:
Follow-up: revert unintended whitespace changes introduced in KAFKA-19891 patch

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -3307,6 +3307,10 @@ private UpdateRegularExpressionsResult maybeUpdateRegularExpressions(
+
+                    if (isNotEmpty(oldSubscribedTopicRegex) && group.numSubscribedMembers(oldSubscribedTopicRegex) != 0) {
+                        updateRegularExpressionsResult = UpdateRegularExpressionsResult.REGEX_UPDATED_AND_RESOLVED;
+                    }
```
We're changing the return value in the wrong branch of the code. We want the branch where `newSubscribedTopicRegex` is empty.

AntonVasant replied:
does this mean the regex itself is empty?

AntonVasant replied:
and not the subscribers counts is empty for the new regex?

squah-confluent replied:
Yes. When we trigger a refresh we do not want to bump the group epoch until the refresh is complete. The refresh will bump the group epoch for us once it has mapped the regex to a list of topics.

AntonVasant replied:
got it, will make the changes accordingly

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -1464,8 +1464,8 @@ public void testStaticMemberGetsBackAssignmentUponRejoin() {
-                    mkTopicAssignment(fooTopicId, 3, 4, 5),
-                    mkTopicAssignment(barTopicId, 2)))
+                mkTopicAssignment(fooTopicId, 3, 4, 5),
+                mkTopicAssignment(barTopicId, 2)))
```
Could we avoid changing the formatting of unrelated tests?

AntonVasant replied:
sure, correcting this

squah-confluent reviewed (commented):
Thanks for the patch! I don't think the fix is in the correct place.

AntonVasant committed:
KAFKA-19899: Return REGEX_UPDATED_AND_RESOLVED when newSubscribedTopicRegex is empty

This patch corrects the return value in
for the case where a member's  becomes empty.

Previously, the branch handling this condition returned , which does not
trigger a group epoch bump during . As a result, the coordinator
could retain stale group metadata when a member unsubscribed from all regex-based topics.

The fix updates the method to return  specifically when the
new subscribed regex string is empty, ensuring that the coordinator clears the member's
assignment and correctly bumps the group epoch.

AntonVasant committed:
KAFKA-19899: Restore tests to original expected behavior after correcting regex empty-branch logic

This commit reverts test modifications that were made under the earlier,
incorrect assumption about the behavior when  becomes empty.
Now that the fix applies to the correct branch, the tests no longer need
their previous adjustments, and their original expectations are valid again.

AntonVasant committed:
KAFKA-19891: Remove trailing whitespace introduced during earlier changes

AntonVasant committed:
KAFKA-19899: Return REGEX_UPDATED_AND_RESOLVED when newSubscribedTopicRegex is empty
This patch corrects the return value in
for the case where a member's  becomes empty.
Previously, the branch handling this condition returned , which does not
trigger a group epoch bump during . As a result, the coordinator
could retain stale group metadata when a member unsubscribed from all regex-based topics.
The fix updates the method to return REGEX_UPDATED_AND_RESOLVED.

dajac commented:
@AntonVasant Could you please add a unit test to cover the change?

AntonVasant committed:
Add test coverage for group epoch bump when member regex becomes empty

AntonVasant committed:
Add test coverage for group epoch bump when member regex becomes empty

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -21009,6 +21009,90 @@ fooTopicName, computeTopicHash(fooTopicName, metadataImage)
     }
+    
+    @Test
+    public void testConsumerGroupMemberJoinsWithNonEmptyRegexToEmptyRegex() {
```
We aren't really joining a new member in this test, only updating an existing member. Maybe we could we rename this to `testConsumerGroupMemberClearsRegex` and move it after `testConsumerGroupMemberJoinRefreshesExpiredRegexes`?

AntonVasant replied:
yeah thats correct

AntonVasant replied:
all the changes have been made as per the comments

squah-confluent replied:
thanks for making the changes. do you mind also doing the reordering I suggested earlier?

AntonVasant replied:
yes, sure

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -21009,6 +21009,90 @@ fooTopicName, computeTopicHash(fooTopicName, metadataImage)
+        Uuid fooTopicId = Uuid.randomUuid();
+        String fooTopicName = "foo";
+        Uuid barTopicId = Uuid.randomUuid();
+        String barTopicName = "bar";
```
Is `bar` used anywhere? We could remove it if not.

AntonVasant replied:
yeah, it doesn't make a difference

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -3307,13 +3307,16 @@ private UpdateRegularExpressionsResult maybeUpdateRegularExpressions(
                     // We also trigger a refresh of the regexes in order to resolve it.
                     throwIfRegularExpressionIsInvalid(updatedMember.subscribedTopicRegex());
                     requireRefresh = true;
+
```
nit: Could we revert this change?

AntonVasant replied:
sure

squah-confluent reviewed (commented):
Thanks for adding the test. Looks good to me. I left some very minor comments.

AntonVasant committed:
Minor cleanup: formatting revert, test rename, unused variable removal

squah-confluent commented:
@AntonVasant Thanks for making the changes. Looks good! Could you respond to my comment here when you have time?

https://github.com/apache/kafka/pull/21013#discussion_r2581329996

AntonVasant committed:
Reorder testConsumerGroupMemberClearsRegex for consistent grouping with regex refresh tests

squah-confluent reviewed (approved):
Thanks for updating the PR!

dajac reviewed (approved):
lgtm, thanks

dajac commented:
Merged to trunk and 4.2.

dajac commented:
I could not cherry-pick it to 4.1 because the code is different in this area. @AntonVasant Would you be willing to raise a PR for the 4.1 branch?

kirktrue commented:
The PR title lists KAFKA-19891, but that Jira is related to logging, which is unrelated to the changes here 🤔

kirktrue commented:
Ah, I think this is KAFKA-19899.

kirktrue commented:
@AntonVasant—for posterity, can you update the PR title to match the correct Jira? Thanks
